### PR TITLE
Avoid NRE in cycle edge case of imperative UDF binding

### PR DESCRIPTION
**Pull request by rick-nguyen**
In scenario of `F():Void = { F(); }`, when finalizing the binding, the Binder will attempt to fetch the CallNode datasource,

Because UDF `TryGetDatasource` override needs the binding, in this cycle/"recursive" edge case, there will be a null reference exception without the check added in this PR.

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests.Shared/RecalcEngineTests.cs [hunk 1]
```diff
@@ -729,6 +729,12 @@ public void DelegatableUDFTest()
             // Imperative function is not delegable
             Assert.True(func.IsAsync);
             Assert.True(!func.IsDelegatable);
+
+            func = recalcEngine.Functions.WithName("E").First() as UserDefinedFunction;
+
+            // Imperative function is not delegable
+            // E():Void = { E() }; ---> binding will be null so no attempt to get datasource should happen
```
[the remark this answers is not in the document]
Comment (rick-nguyen) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests.Shared/RecalcEngineTests.cs:736]: It's a valid binding, the error is added during cycle detection in analysis

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:
